design/animation.md

```diff
@@ -0,0 +1,317 @@
+Animations
+===========
+
+Our current animation system is under-featured, unstable, and based on tech (Unity Animation components) that is too
+rigid for our purposes. We need an animation system that can support the following:
+
+1. Native animations - Animations added to actors from glTF or native Unity sources should support the basic API: play,
+	pause, seek.
+2. Keyframes - Vary actor/asset fields through a sequence of values on a timeline.
+	1. Interpolation - Make simple single-target two-frame animations easy to generate.
+	2. Moving targets - Setting/interpolating to/from another actor's dynamic values. Useful for look-at and
+		fly-to type animations.
+3. Easing - Support different interpolation curves between keyframes.
+	1. No easing - Setting fields that cannot be interpolated should also be supported (string, boolean, etc.).
+4. Chaining - Begin an animation immediately after another finishes, or on some other event (behavior tie-in?)
+5. Multiple targets - Animate multiple actors/assets, and multiple properties, in lock-step (equivalent to Unity's
+	animation curves)
+6. Blending - Allow multiple animations to run on the same target simultaneously, and blend the computed destinations
+	into one, proportional to a weight property.
+
+To address this need, I propose the following API:
+
+API Design
+-----------
+
+Animations are assets. They are created explicitly by a function on asset containers (
+`AssetContainer.createAnimation`). Native animations from library or glTF actors do not create these assets, instead
+creating animation instances only (see [Instances](#instances)).
+
+When created directly, animations are created as an array of "tracks", each composed of:
+1. A generic path to a field that can be animated (see [Paths](#paths)).
+2. An array of keyframes, each composed of:
+	1. A time value.
+	2. A valid value for the targeted path.
+	3. An optional bezier easing function.
+
+```ts
+type Animatible = Actor | Material;
+
+class Animation {
+	id: Readonly<string>;
+	duration: Readonly<number>;
+	targetCount: Readonly<number>;
+}
+
+class AssetContainer {
+	public createAnimation(name: string, options: { targetCount?: number, tracks: Track<AnimationProp>[] }): Animation;
+}
+
+type AnimationProp = Vector2 | Vector3 | Quaternion | Color3 | Color4 | number | string | boolean;
+
+type Track<T extends AnimationProp> = {
+	target: TargetPath<T>;
+	keyframes: Keyframe<T>[];
+}
+
+type Keyframe<T extends AnimationProp> = {
+	time: number;
+	value: T;
+	easing?: EasingFunction;
+}
+
+type EasingFunction = [number, number, number, number];
+```
+
+### Paths
```

Comment:
Can you add an example of interpolating one actor to another actor's position?

Reply:
I'd like to see the syntax for referencing the field of another actor as a target value.

Reply:
Whoops, I totally overlooked that use case. Will update.

---
In reply to: [341698633](https://github.com/microsoft/mixed-reality-extension-sdk/pull/464#discussion_r341698633) [](ancestors = 341698633)